src/clj/puppetlabs/jdbc_util/cp.clj

```diff
@@ -0,0 +1,183 @@
+   values."
+  [config :- HikariConfig
+   properties :- {s/Str s/Str}]
+  (doseq [[k v] properties]
+    (add-datasource-property config k v)))
+
+(defn validate-options
+  "Manage defaulting and validation of options."
+  [options]
+  (s/validate ConfigurationOptions (merge default-datasource-options options)))
+
+(s/defn ^:always-validate datasource-config :- HikariConfig
+  "Produce a valid HikariConfig object from provided configuration options."
+  [datasource-options]
+  (let [config (HikariConfig.)
+        options (validate-options datasource-options)
+        {:keys [adapter
+                auto-commit
+                connection-test-query
+                connection-timeout
+                idle-timeout
+                initialization-fail-fast
+                internal-test-query
+                isolate-internal-queries
+                max-lifetime
+                maximum-pool-size
+                metrics-registry
+                minimum-idle
+                password
+                pool-name
+                read-only
+                register-mbeans
+                stats
+                subname
+                subprotocol
+                user
+                validation-timeout]} options]
+
+    ;; Set pool-specific properties
+    (doto config
+      (.setAutoCommit auto-commit)
+      (.setInitializationFailFast initialization-fail-fast)
+      (.setRegisterMbeans register-mbeans)
+      (.setJdbcUrl (str "jdbc:" subprotocol ":" subname))
+      (.setIsolateInternalQueries isolate-internal-queries)
+      (.setConnectionTestQuery connection-test-query)
+      (.setPoolName pool-name))
+
+    ;; Set optional properties
+    (when stats (.setMetricRegistry config metrics-registry))
+    (when read-only (.setReadOnly config read-only))
+    (when idle-timeout (.setIdleTimeout config idle-timeout))
+    (when max-lifetime (.setMaxLifetime config max-lifetime))
+    (when minimum-idle (.setMinimumIdle config minimum-idle))
+    (when maximum-pool-size (.setMaximumPoolSize config maximum-pool-size))
+    (when validation-timeout (.setValidationTimeout config validation-timeout))
+    (when connection-timeout (.setConnectionTimeout config connection-timeout))
+    (when user (.setUsername config user))
+    (when password (.setPassword config password))
+
+    ;; Set driver specific properties
+    (case subprotocol
+      "postgresql" (add-datasource-properties config {"ApplicationName" pool-name
+                                                      "tcpKeepAlive" "true"
+                                                      "logUnclosedConnections" "true"})
+      "default")
+
+    config))
+
+(defn pooled-db-spec
+  "Create a pooled clojure.java.jdbc compatible db-spec."
+  [pool-name options]
+  (let [options (assoc options :pool-name pool-name)
+        config (datasource-config options)
+        orig-ds (HikariDataSource. config)
+        jdbcdslog (ConnectionPoolDataSourceProxy.)
```

Comment:
Here we always proxy via the statement logger, we can switch this off though with log-statements so I'll need to make this optional.